Certificate role based authorization in Apache Bookkeeper.

Ghatage opened the pull request:
This feature allows a predefined set of services to be 'whitelisted' to be able
to access bookkeeper based on their client certificates.

### Motivation
As BookKeeper and its supported services move to a cloud friendly service based architecture, it becomes of utmost importance to monitor and allow only certain qualified services to be able to access the data in BK.
We have TLS based authentication, however, any service with the rootCA can access Bookkeeper clusters which is not desirable.


### Changes
So this feature can be broken down into two parts:
1. Certificate and roles
2. Server configuration for authorized roles

Details:
1. Certificate and roles:
Here is an example of how the SUBJECT field of a final certificate for Apache Pulsar running in the cloud would look like:
    CN=apache.bookkeeper.org
O=apache-pulsar
OU=0:pulsar-broker-role;1:cluster-1
L=San Francisco
S=CA
C=US
This shows that this bookkeeper client certificate is owned by the apache pulsar service has the role ‘pulsar-broker-role’ for entities in ‘cluster-1’.
Only those services with pulsar-broker-role should be able to access it.
We can add more fields separated by commas to increase the upstream application clusters to be able to access this bookkeeper cluster.

For example: OU=0:herddb-readonlyNode,herddb-readwriteNode;1:herddb-cluster2

Such separation of access based on services is paramount to keeping this secure as many upstream users of BookKeeper are financial institutions, databases and other services.

2. Server configuration for authorized roles
Once we have a certificate whose SUBJECT field has the OU attribute with the roles we want to authorize, on the Bookie side, we need to specify which roles are allowed.
We make this happen by introducing a server configuration option called ‘authorizedRoles’.
Since we have only static options, this will be set in stone as long as the bookie booted up with it.
If in case we need to change the allowed roles, we’ll need to stop the bookie, update the configuration file and then restart the bookie.
We can have multiple roles which are authorized as the OU field can have multiple comma separated values for roles.

Master Issue: #2354

eolivelli commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/tls/BookieAuthZFactory.java:
```diff
@@ -0,0 +1,133 @@
+ * Authorization factory class.
+ */
+@Slf4j
+public class BookieAuthZFactory implements BookieAuthProvider.Factory {
```
what about BookieX509CertificateAuthZFactory ?

Ghatage replied:
I don't know about this class, I checked the code and it doesn't seem to be there..

eolivelli commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/tls/BookieAuthZFactory.java:
```diff
@@ -0,0 +1,133 @@
+
+        if (allowedRoles == null || allowedRoles.length == 0) {
+            throw new RuntimeException("Configuration option \'bookieAuthProviderFactoryClass\' is set to"
+                    + " \'BookieAuthZFactory\' but no roles set for configuration field \'authorizedRoles\'.");
```
what about usiing this.class.getName() instead of an hardcoded "BookieAuthZFactory" ?

Ghatage replied:
Gonna let this be, as the message is specific to this class.

eolivelli commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/util/CertUtils.java:
```diff
@@ -0,0 +1,117 @@
+/**
+ * Certificate parsing utilities.
+ */
+public class CertUtils {
```
abstract ? 
with a private default constructor

Ghatage replied:
Fixed in [7276947](https://github.com/apache/bookkeeper/commit/72769477f52e4982081ae87a8e43da0a0e34e2d1)

eolivelli commented on bookkeeper-server/src/test/java/org/apache/bookkeeper/tls/TestTLS.java:
```diff
@@ -635,6 +635,47 @@ public void testBookieAuthPluginRequireClientTLSAuthenticationLocal() throws Exc
 
+    /**
+     * Verify that given role in client certificate is checked when BookieAuthZFactory is set.
+     * Positive test case where all given roles are present.
```
this test does look like a "Positive test"
you are asserting that the client is not able to work

Ghatage replied:
If `testClient()` fails, it will be because of an exception thrown during Authorization.
And hence we catch that and fail the test.
Otherwise, if the test passes, it never reaches the `catch` block and finishes, passing the test
Sorry for the confusing comment, I can change that in a follow up.

Ghatage replied:
Fixed in [7276947](https://github.com/apache/bookkeeper/commit/72769477f52e4982081ae87a8e43da0a0e34e2d1)

eolivelli commented on conf/bk_server.conf:
```diff
@@ -226,6 +226,15 @@ httpServerClass=org.apache.bookkeeper.http.vertx.VertxHttpServer
+# comma separated values of roles from the certificates OU field which we want to authorize for access
+#
+# Example settings:
+#   authorizedRoles=apache-pulsar-cluster-12,apache-kafka-cluster-2
```
looks like 'kafka' is not useful in BK context :-)

Ghatage replied:
Fixed in [7276947](https://github.com/apache/bookkeeper/commit/72769477f52e4982081ae87a8e43da0a0e34e2d1)

eolivelli commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieRequestProcessor.java:
```diff
@@ -564,10 +564,19 @@ public void operationComplete(Future<Channel> future) throws Exception {
+                     * Success of the future doesn't guarantee success in authentication
+                     * future.isSuccess() only checks if the result field is not null
+                     */
+                    if (future.isSuccess() && authHandler.authenticated) {
```
can we add a "isAuthenticated" method to authHandler ?
this way we want access directly the field

Ghatage replied:
Fixed in [7276947](https://github.com/apache/bookkeeper/commit/72769477f52e4982081ae87a8e43da0a0e34e2d1)

lamber-ken commented:
rerun failure checks

rdhabalia commented:
rerun failure checks

rdhabalia commented:
@Ghatage do we need this patch for 4.11?

Ghatage commented:
> @Ghatage do we need this patch for 4.11?

@rdhabalia No, we can leave this for later.

Ghatage commented:
Hey @eolivelli,

I've made the following changes as requested:
* Added a test class for `BookieAuthZFactory` which doesn't spin up a bookie cluster.
* Added all negative test cases to the above new test class.
* Removed the 'heavy' negative test case from `TestTLS` class.
* Modified comment to explain the success case better.
* Converted `CertUtils` class to abstract with a private default constructor.
* Removed unrelated comments from `bk_server.conf`
* Added `isAuthenticated()` method to authProvider

Feature wise this is complete and switched off by default.
Lastly, once the CI tests pass, it's best to hold off on merging this before 4.11 is cut and released as discussed in the mailing list.

Ghatage commented:
rerun failure checks

Ghatage commented:
rerun failure checks

Ghatage commented:
rerun failure checks

Ghatage commented:
The test suite runs fine locally and the test I've written doesn't fail either, two other tests randomly fail with `Not Enough Bookies`.
The option set is only for that test and by default it's false.
@sijie, @eolivelli  any idea why `TestTLS` keeps failing?

Ghatage commented:
Messed up the PR, closing this and reopening under new PR.

Ghatage commented:
Reopened under https://github.com/apache/bookkeeper/pull/2429
